lib/mix/tasks/firmware.ex

```diff
@@ -198,4 +200,44 @@ defmodule Mix.Tasks.Firmware do
     less_than_elixir_19 = Nerves.elixir_version() |> Version.compare("1.9.0") == :lt
     less_than_elixir_19 && Code.ensure_loaded?(Mix.Tasks.Distillery.Release)
   end
+
+  defp compiler_check() do
+    with {:ok, otpc} <- compiler_version(:code),
+         {:ok, otp_requirement} <- Version.parse_requirement("~> #{otpc.major}.#{otpc.minor}"),
+         {:ok, elixirc} <- compiler_version(:"Elixir.Kernel") do
+      unless Version.match?(elixirc, otp_requirement) do
+        Mix.raise("""
+        The Erlang compiler that compiled Elixir is older than the compiler
+        used to compile OTP.
+
+        Elixir: #{elixirc.major}.#{elixirc.minor}.#{elixirc.patch}
+        OTP:    #{otpc.major}.#{otpc.minor}.#{otpc.patch}
+
+        Please use a version of Elixir that was compiled using the same major
+        version of OTP.
+
+        For example:
+
+        If your target is running OTP 22, you should use a version of Elixir
+        that was compiled using OTP 22.
```

Comment:
How about adding this right after this sentence? I think the hardest part is knowing that Elixir is published as a precompiled binary against several OTP versions.

```
If you're using asdf to manage Elixir versions, run:

asdf install elixir #{System.version()}-otp-#{otpc.major}
asdf global elixir #{System.version()}-otp-#{otpc.major}
```